### ISSUE TITLE
An EntityStruct declares entities too, scoped as its children

### ISSUE THREAD

**Issue by sunarya-thito** [sunarya-thito/good#90]
A `SceneStruct` declares the entities a scene can spawn. An `EntityStruct`
should be able to declare entities too — and because of where the declaration
sits, they are children of the declaring entity:

```dart
class Galaxy extends SceneStruct {
  final mote = EntityStruct.of<Mote>();      // a scene entity
}

class Turret extends EntityStruct with Parent {
  final barrel = EntityStruct.of<Barrel>();  // a child of every Turret
}
```

Same call, same spelling, one rule: **the scope of the declaration says what
kind of entity it is.** Spawning a `Turret` spawns a `Barrel`, links it in as a
child, and destroying the `Turret` takes it with it.

There is no new concept and no new vocabulary. This is entity declaration,
scoped.

## Why this is not blocked, though #57 said it was

#57 left `EntityStruct.of<T>()` unimplemented and gave a lifecycle reason: a
declaration needs a `MemoryPool`, and a scene only gets one at
`initializeScene`, which is public precisely so tests can build scenes by hand.

**That is true of `SceneStruct` and `Game`, and false of `EntityStruct`:**

- `SceneStruct.initializeScene` (`scene.dart:267`) assigns `_pool = pool` and
  *then* calls `describeScene(_SceneDescriptor(...))`.
- `_SceneDescriptor.has` (`scene.dart:487`) calls
  `ArchetypeRegistry.reserve(_scene.pool)`, pushes the data descriptor, and
  *then* runs `object = create()`.

So a prefab's field initialisers already run with a live pool and an open
descriptor. A nested `EntityStruct.of<Barrel>()` inside `Turret`'s initialisers
runs during that `create()`. The chicken-and-egg is at the scene and game level
only, which is exactly where this issue does not need it.

The stack in `packages/good/lib/src/declare.dart` is a stack rather than a
single slot for this shape — #57 verified three levels of nesting in one
expression.

## The open question: what it returns

`EntityStruct.of<T>()` returns `T` at scene scope, which is what makes
`mote.someField[entity]` resolve. #57 verified a generic *class* cannot do this,
since constructors cannot have type parameters.

At child scope two things are wanted, and they are not the same object:

1. the child's **prefab**, to read its columns; and
2. the child **`Entity` for a given parent entity** — which has to be a column
   on the parent's row. The declaration runs once per archetype, not once per
   entity, so a field holding an `Entity` would be one value shared by every
   `Turret`. That is the bug `docs/guide/entities-and-components.md` opens its
   `!!! danger` block with.

Whoever implements this decides how both are reached **without `of<T>()`
meaning two different things depending on where it appears**. One lean, not a
decision: keep the return type `T` in both scopes, auto-declare the handle
column behind it, and reach the child through the accessor API from `a462113` —
in the shape of `turret<Parent>().child(barrel)`. A different return type at
child scope is defensible *provided the rule fits in one sentence*. If it does
not, the design is wrong.

## The hook, if one is needed, goes on `Parent`

If the unified form proves genuinely blocked here too, the fallback is a
`describeChildren` hook — and it is declared on **`Parent`**, not on
`EntityStruct`:

```dart
mixin Parent on Component {
  @mustCallSuper
  void describeChildren(ChildrenDescriptor descriptor) {}
  ...
}
```

used as

```dart
@override
@mustCallSuper
void describeChildren(ChildrenDescriptor desc) {
  super.describeChildren(desc);
  barrel = desc.has(Barrel.new);
}
```

That is the established shape for a component-scoped declaration hook, not a new
one: `describeCollider` is declared on `mixin Collider2D on MultiComponent`
(`collider.dart:479`) and `describeSprites` on `Renderable2D`
(`render_2d.dart:719`), both `@mustCallSuper` with an empty body, both
overridden by the concrete prefab.

It also **turns the `Parent` requirement into a compile-time one**. A struct that
has not mixed in `Parent` has no `describeChildren` to override, so declaring
children without being able to hold them stops being a registration-time error
message and becomes something that does not compile.

The pass is `if (object is Parent) object.describeChildren(...)` in
`_SceneDescriptor.has`, alongside `describeType` / `describeAssets` /
`describeStruct`. Note `describeAnimation` next to it is called unconditionally
and the comment says why — every `EntityStruct` has `Animations`. `Parent` is
not universal, so this one is guarded.

Note also that `desc.has(Barrel.new)` is already the same spelling
`SceneDescriptor` takes since #57, so even the fallback keeps the point.

## What has to be true

- **The declarer mixes in `Parent`, the child prefab mixes in `Child`.** Both
  are enforced at runtime today by `addChild` and `addEntityIn`. The `Parent`
  half should be compile-time — see the hook section below. The `Child` half
  cannot be, since the declaration names a type rather than a value the analyzer
  can constrain, so it fails at *declare* time: it is a static fact about the
  prefab and there is no reason to wait for a spawn to report it.
- **The child spawns in the parent's scene**, at mount, already linked.
  `SceneStruct.addEntity(prefab, parent: ...)` creates the child and calls
  `addChild` itself.
- **They nest.** A `Barrel` that declares its own children mounts them too.
- **A cycle is a declare-time error.** `Turret` declaring a `Turret` is
  unbounded recursion at mount. The check belongs at registration, where the
  whole graph is known, and the message should name the cycle.
- **Teardown needs no new code, but does need a test.** `Entity.destroy()`
  (`scene.dart:645`) already destroys children first, recursively, and
  `hierarchy_test.dart:381` proves a three-level subtree's rows all come back. A
  declared child is an ordinary child: it goes with its parent and is not
  promoted. Assert that rather than assuming it.

## Two hazards, both already documented in this repo

**Mid-tick allocation is an open bug.** #5 — rows allocated mid-tick lose their
writes; the write slot gets the right value and the published snapshot does not.
Mounting a child *is* a mid-tick allocation followed immediately by writes to
both rows. Probe it before building on it, and if it reproduces, report the case
on #5 — a second independent reproduction is worth more than a guess at a cause.

**Several children in one tick is the shape that was already broken once.**
`addChild` reads `lastChild` through `readPending` and not an ordinary read, and
the comment above it in `hierarchy.dart` says why: two same-tick `addChild`
calls otherwise both see the same stale published tail, both take the
`oldLast == null` branch, and the second silently overwrites the first, leaving
the parent one child and orphaning the rest with no error. It names "spawning a
character with three attachments from one command" as the case. **A prefab
declaring three children is that case by construction, every time.** Test three,
or ship the bug the comment was written about.

**Done when** an `EntityStruct` declares entities the same way a `SceneStruct`
does, spawning the parent spawns them linked, destroying the parent destroys
them, and a prefab declaring three children keeps all three.

**Comment by sunarya-thito** [comment]
Landed in `57f8878`. Suites verified independently by me: `good` **529** (507 + 22 new), `goo2d` 221, `goo3d` 37, `goo2d_physics_box2d` 64, `good_cli` 162, example 25 (+1 skipped).

```dart
class Turret extends EntityStruct with Transform2D, Parent {
  final barrel = EntityStruct.of(Barrel.new);
}

final barrelEntity = turretEntity<Parent>()[turret.barrel];
```

## Where this issue was wrong

**`EntityStruct.of<Barrel>()` is impossible, and not for the reason this issue argued about.** The issue debated whether the *lifecycle* blocker from #57 applied at entity scope. It does not — that part was right. But the spelling fails earlier, on a language rule: Dart has no `new T()`, so `T()` is `invocation_of_non_function` (verified on 3.13.0). A type argument alone cannot construct the prefab and there is no factory registry to look one up in — #18 would be that registry. So it is the tear-off, `EntityStruct.of(Barrel.new)`, identical to what `descriptor.has` has taken since #57 and for the same reason.

**The `describeChildren`-on-`Parent` fallback was not needed**, so the compile-time `Parent` guarantee I argued for it never materialised. Declaring children without `Parent` is a registration-time `StateError`. What *is* static is the read: `operator []` takes a `Child`, not an `EntityStruct` as I specified — which accepts every prefab that could legally have been declared and makes the rest a compile error rather than a runtime one. That is stricter than I asked for and better.

## #5 reproduced on this path

Posted to #5 with the numbers. On a **recycled** row, `allocateRow` stamps defaults into the write slot *and* the published slot, so the mount's write is invisible to an ordinary read for the rest of that tick. On a `hasEntity` column the wrong answer would be `Entity(0)` — a real, resolvable handle naming somebody else's row, not a detectable null. Hence `optEntity` and `readPending`. The committed test asserts current #5 behaviour explicitly, so a fix to #5 fails there naming the assumption that changed.

## `EntityLifecycleListener`: half works, and the half that does not is decisive

A `Component` mixin **does** receive the dispatch — that is a committed test, and it answers the question I asked. Mounting is still a guarded walk in `addEntityIn` anyway, for two reasons found by testing rather than argument:

- an override calling `super.onEntityMounted` **last** runs its own code before the children exist, so the ordering guarantee becomes the user's to keep;
- an override that **omits** `super` silently suppresses the hook — declared children never spawn, no error. Exactly the failure mode the listener's own doc warns about, reached by a different mechanism than expected.

Teardown is worse: `unmountEntitiesOf` fires the same dispatcher for every row of an unloading scene, so a destroy cascade in `onEntityUnmounted` would free rows the unload is already freeing. `Entity.destroy()`'s `tryGet` branches stayed. The guard is one `if (prefab is Parent)` in a method that already asks the same question twice.

## Four hierarchy operations, not two

`removeChild` becoming destructive left two holes, so:

| | |
|---|---|
| `parent<Parent>().addChild(child)` | attach; **throws** if already parented |
| `parent<Parent>().adopt(child)` | move it here from wherever it is |
| `parent<Parent>().removeChild(child)` | destroys child and subtree |
| `child<Child>().detach()` | unlink, stay alive as a root |

`detach` is not spare surface: two existing tests unparent an entity and keep it alive to exercise the childless fast path, and neither destroy nor reparent expresses that.

`Entity.destroy()` calling `removeChild` was infinite recursion once `removeChild` destroyed — caught by a stack overflow in the suite, now routed through a shared `unlinkChild` splice. Worth recording as the kind of thing that only a test finds.

Cycles are rejected where the link is made, by walking up from the prospective parent; self-parenting and ancestor-into-descendant are both tested. Not a depth cap — `destroy()` recurses a subtree with no cap and would simply not terminate.

**Comment by sunarya-thito** [comment]
Landed in `57f8878`. Suites: `good` **529** (507 + 22 new), `goo2d` 221, `goo3d` 37, `goo2d_physics_box2d` 64, `good_cli` 162, example 25 (+1 skipped).

```dart
class Turret extends EntityStruct with Transform2D, Parent {
  final barrel = EntityStruct.of(Barrel.new);
}

final barrelEntity = turretEntity<Parent>()[turret.barrel];
```

## Where this issue was wrong

**`EntityStruct.of<Barrel>()` is impossible, and not for the reason argued above.** The debate was whether the *lifecycle* blocker from #57 applied at entity scope. It does not — that part was right. But the spelling fails earlier, on a language rule: Dart has no `new T()`, so `T()` is `invocation_of_non_function` (verified on 3.13.0). A type argument alone cannot construct the prefab and there is no factory registry to look one up in — #18 would be that registry. So it is the tear-off, `EntityStruct.of(Barrel.new)`, identical to what `descriptor.has` has taken since #57 and for the same reason.

**The `describeChildren`-on-`Parent` fallback was not needed**, so the compile-time `Parent` guarantee argued for it never materialised. Declaring children without `Parent` is a registration-time `StateError`. What *is* static is the read: `operator []` takes a `Child`, which accepts every prefab that could legally have been declared and makes the rest a compile error rather than a runtime one — stricter than this issue asked for.

## #5 reproduced on this path

Posted to #5 with numbers. On a **recycled** row, `allocateRow` stamps defaults into the write slot *and* the published slot, so the mount's write is invisible to an ordinary read for the rest of that tick. On a `hasEntity` column the wrong answer would be `Entity(0)` — a real, resolvable handle naming somebody else's row, not a detectable null. Hence `optEntity` and `readPending`. The committed test asserts current #5 behaviour explicitly, so a fix to #5 fails there naming the assumption that changed.

## `EntityLifecycleListener`: half works, and the failing half decides it

A `Component` mixin **does** receive the dispatch — committed as a test. Mounting is still a guarded walk in `addEntityIn` anyway, for two reasons found by testing rather than argument:

- an override calling `super.onEntityMounted` **last** runs its own code before the children exist, so the ordering guarantee becomes the user's to keep;
- an override that **omits** `super` silently suppresses the hook — declared children never spawn, no error. Exactly the failure mode the listener's own doc warns about, reached by a different mechanism than expected.

Teardown is worse: `unmountEntitiesOf` fires the same dispatcher for every row of an unloading scene, so a destroy cascade in `onEntityUnmounted` would free rows the unload is already freeing. `Entity.destroy()`'s `tryGet` branches stayed. The guard is one `if (prefab is Parent)` in a method that already asks the same question twice.

## Four hierarchy operations, not two

| | |
|---|---|
| `parent<Parent>().addChild(child)` | attach; **throws** if already parented |
| `parent<Parent>().adopt(child)` | move it here from wherever it is |
| `parent<Parent>().removeChild(child)` | destroys child and subtree |
| `child<Child>().detach()` | unlink, stay alive as a root |

`detach` is not spare surface: two existing tests unparent an entity and keep it alive to exercise the childless fast path, and neither destroy nor reparent expresses that.

`Entity.destroy()` calling `removeChild` was infinite recursion once `removeChild` destroyed — caught by a stack overflow in the suite, now routed through a shared `unlinkChild` splice.

Cycles are rejected where the link is made, by walking up from the prospective parent; self-parenting and ancestor-into-descendant are both tested. Not a depth cap — `destroy()` recurses a subtree with no cap and would simply not terminate.
